### PR TITLE
Enable CUDA builds

### PR DESCRIPTION
**Pull request by basnijholt**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by basnijholt** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml
Comment (leofang) [inline recipe/meta.yaml:31]: This should not be set, because for most of conda-forge packages we want to link dynamically.
```suggestion
```

File: recipe/meta.yaml
Comment (leofang) [inline recipe/meta.yaml:53]: @wshanks I don't think Qiskit Aer depends on cuQuantum Python yet?
```suggestion
```

File: recipe/meta.yaml
Comment (leofang) [inline recipe/meta.yaml:60]: ditto
```suggestion
```

File: recipe/meta.yaml
Comment (leofang) [inline recipe/meta.yaml:20]: This skip condition needs an update with the presence of CUDA (since cuQuantum doesn't support Windows), let's refine it:
```suggestion
  skip: true  # [not linux64]
```
We may need to further distinguish CUDA 11/12, but it can be done at a later time when the bot sends a migration PR to this feedstock.

**Review by leofang — CHANGES_REQUESTED** [review]
My apology for long silence @wshanks @basnijholt. Sorry I wasn't able to catch up with OSS requests lately. I've made some suggestions below. Please do let me know there's any other matter that I should attend to, thanks 🙂

File: recipe/meta.yaml
Comment (leofang) [reply recipe/meta.yaml:20]: (updated the skip condition a bit)

Comment (leofang) [reply recipe/meta.yaml:20]: Note that while this is effectively done at the dependency level, we still need to be explicit here so as to reduce the number of Windows CIs (see the excessive number of CI pipelines launched).

File: recipe/meta.yaml
Comment (wshanks) [reply recipe/meta.yaml:53]: I think cuquantum support was added in [this PR](https://github.com/Qiskit/qiskit-aer/pull/1854). It adds some extra dependencies in setup.py when building the qiskit-aer-gpu package for PyPI. I have not researched this much yet. I think those dependencies it lists are just libraries needed by the C++ code (but hosted on PyPI) and are separate from cuquantum-python. I don't see a reference to `cuquantum` in the Python code of qiskit-aer.

**Comment by wshanks** [comment]
@leofang I have not looked closely. Is the approach here in keeping with what you outlined in #5  and your concerns about https://github.com/Qiskit/qiskit-aer/issues/1686?

**Comment by leofang** [comment]
> Is the approach here in keeping with what you outlined in https://github.com/conda-forge/qiskit-aer-feedstock/issues/5 and your concerns about https://github.com/Qiskit/qiskit-aer/issues/1686?

Thanks for reminder, Will. I am slowly recollecting my memory. 

You're right that we cannot yet build an all-in-one package, as it'd impact CPU-only users (even if they have GPU installed). We need to use the old CPU/GPU approach, as done in the old UCX package (which doesn't do it anymore) or the PyTorch package (we still do this on conda-forge), meaning we need to generate multiple outputs here + add a mechanism to avoid CPU/GPU packages from co-existing. It's doable but not fun.

If it's fixed then making an all-in-one package might become simpler, but let me double check and ensure I don't miss other corner cases.

**Comment by leofang** [comment]
@conda-forge-admin, please rerender

**Comment by leofang** [comment]
@conda-forge-admin, please rerender

**Comment by leofang** [comment]
In one of the CI logs I see
```
2023-12-12T03:32:11.0337993Z   nvcc error   : 'cicc' died due to signal 9 (Kill signal)
```
which is usually a sign of OOM (the VMs used here are not beefy unfortunately...). Let's see how we can reduce stressing the machines...

**Comment by basnijholt** [comment]
Great, that seems to have worked 🥳

**Comment by leofang** [comment]
Not sure what's going on, I need to take a closer look first. btw @wshanks does Aer still support CUDA 11?

**Comment by leofang** [comment]
@conda-forge-admin, please rerender

**Comment by basnijholt** [comment]
How could you tell it's disk space? I don't see any error messages in the last jobs at least.

**Comment by leofang** [comment]
> How could you tell it's disk space? I don't see any error messages in the last jobs at least.

By experience? LOL... 

I still think it's OOM, so I wanted to increase the swap size (to reduce memory pressure by swapping to disk), but to enable swapping we need more disk space too, and with CUDA sw stack I had prior experience in hitting out of space, so I figured I also better clean up disk as a precaution. 

Anyway, it builds now, let me fix the overlooking error.

**Comment by leofang** [comment]
> Anyway, it builds now, let me fix the ~~overlooking~~ overlinking error.

Tracked in https://github.com/conda-forge/cuquantum-feedstock/issues/53.

**Comment by basnijholt** [comment]
This is awesome @leofang! Thanks a lot for fixing the builds! :tada: 😄

I just added build strings such that `qiskit-aer` can be installed using `qiskit-aer * cpu*` or `qiskit-aer * cuda*`.

Anything we still need before we can merge this?

**Comment by leofang** [comment]
> Anything we still need before we can merge this?

Yes this is precisely the next issue I plan to discuss. What's done here (make two builds distinguishable) is not enough. We need 2 extra steps:
1. Prioritize one of the builds (either CPU or GPU, I don't care, either of you should chime in)
2. Make sure two builds cannot coexist  

Now, there are multiple ways to make two builds distinguishable, and each approach will have its own way for users to select which build they'd like to install:
- using build strings (currently done)
  - Ex: `conda install qiskit-aer=*=cuda*`
- using a mutex meta package and/or conflicting `run_constrained` requirement 
  - Ex: `conda install qiskit-aer qiskit-aer-proc=*=cuda`, or
  - Ex: `conda install qiskit-aer cuda-version=X.Y`
- using different package names
  - Ex: `conda install qiskit-aer-cuda`

The point is if a future change is possible to allow an all-in-one build, the installation instruction may change. @wshanks what's your thought on this?

File: recipe/meta.yaml
Comment (leofang) [reply recipe/meta.yaml:69]: @basnijholt I guess so, could you open a PR and we can discuss there?

**Comment by basnijholt** [comment]
I went with `=*=cuda*` in qsimcirq-feedstock because I found that almost all other packages on `conda-forge` do that now.

The following packages _**converted**_ from the `package-name-cuda` format to build strings (non exhaustive list)

- https://github.com/conda-forge/kaldi-feedstock/blob/f0e1fb122f079b27af9f19a78d100f192e4c35f9/recipe/meta.yaml#L57C1-L58
- https://github.com/conda-forge/pytorch-cpu-feedstock/blob/7c7a57b7515eaeda67d3879b56b68466f38f0b0d/recipe/meta.yaml#L54-L55
- https://github.com/conda-forge/tensorflow-feedstock/blob/f0f57d5b35b10014ea9cbfe184472c021a3863e7/recipe/meta.yaml#L419C10-L420
- https://github.com/conda-forge/jaxlib-feedstock/blob/2ffd94498112c6e352aef3fe58856b9281697091/recipe/meta.yaml#L25-L26
- https://github.com/conda-forge/torchvision-feedstock/blob/cf89794338c46f3dd4cee49930a666d6e8ea5290/recipe/meta.yaml#L43-L44

And many other packages that used to have a split package (`package-name-cuda`) do not have that anymore:
- https://github.com/conda-forge/xgboost-feedstock/blob/main/recipe/meta.yaml
- https://github.com/conda-forge/arrow-cpp-feedstock/blob/main/recipe/meta.yaml
- https://github.com/conda-forge/numba-feedstock/blob/main/recipe/meta.yaml


I can find more examples but I think it is quite clear 😄

**Comment by basnijholt** [comment]
Can we not do both the build string and add the conflicting `run_constrained` requirement?

**Comment by basnijholt** [comment]
> Prioritize one of the builds (either CPU or GPU, I don't care, either of you should chime in)

I now bump the build number for CUDA builds just like is done for [pytorch](https://github.com/conda-forge/pytorch-cpu-feedstock/blob/7c7a57b7515eaeda67d3879b56b68466f38f0b0d/recipe/meta.yaml#L5-L9).

> Make sure two builds cannot coexist

Not sure how to do this.

**Comment by leofang** [comment]
Thanks, Bas.

> > Make sure two builds cannot coexist
> 
> Not sure how to do this.

Forget about this, I was probably thinking something else, sorry.

> I can find more examples but I think it is quite clear that it's not the split package solution 😄

Sure, but this only rules out Option No.3. I don't like it either, but I'd like to have @wshanks involved in the technical discussion since he's the main maintainer. 

We still have Option 2 on the table, which does not require changing the build string of `qiskit-aer`. The way a conflict can be made (thus allowing selecting a build) is to do
```yaml
- run_constrained:
  -  qiskit-aer-proc * cuda  # [cuda_compiler_version != "None"]
  -  qiskit-aer-proc * cpu   # [cuda_compiler_version == "None"]
```
if one also generates a new output `qiskit-aer-proc` with a build string "cpu" or "cuda" in this feedstock. Alternatively, one can use the CUDA package selector as the mutex:
```yaml
- run_constrained:
  -  cuda-version <0.a0  # [cuda_compiler_version == "None"]
```
in which case the CPU build would not be installable to a conda env with CUDA packages present, and depending on the use case one may or may not like it.

File: recipe/meta.yaml
Comment (leofang) [inline recipe/meta.yaml:13]: nit:
```suggestion
{% if cuda_compiler_version in (None, "None", True, False) %}
{% set cuda_major = 0 %}
{% else %}
{% set cuda_major = environ.get("cuda_compiler_version", "11.2").split(".")[0] | int %}
{% set build = build + 200 %}
{% endif %}
```

File: recipe/meta.yaml
Comment (leofang) [inline recipe/meta.yaml:50]: nit:
```suggestion
    - {{ compiler('cuda') }}  # [linux64 and cuda_compiler_version != "None"]
```

**Review by leofang — APPROVED** [review]

**Review by leofang — APPROVED** [review]
Thanks, Bas. LGTM. Let's wait for another day and see if Will has time to chime in.

**Comment by wshanks** [comment]
> btw @wshanks does Aer still support CUDA 11?

Yes, it is still referenced in the documentation and code, though so far they did not publish gpu wheels for the most recent qiskit-aer release.

I am not that familiar with cuda builds, but from what I understand it sounds like if https://github.com/Qiskit/qiskit-aer/issues/1686 were fixed we could make an all-in-one build? So that would mean we would have to stop modifying the build string. I don't think making cpu version incompatible with the cuda libraries is that disruptive of a constraint. So

```
- run_constrained:
  -  cuda-version <0.a0  # [cuda_compiler_version == "None"]
```

seems okay to me, assuming that installs the cpu version on a system without a gpu (rather than install the cuda libraries as well onto it).

The build string approach seems okay as well. We can just live with the disruption when the all-in-one build becomes possible. Will it default to the cpu version? I think that would be the best default, assuming conda has no way specify a preference, like install the gpu version if cuda-version is already installed. I would just want to avoid `conda install qiskit-aer` installing the gpu version on a system without a gpu.

**Comment by leofang** [comment]
> I am not that familiar with cuda builds, but from what I understand it sounds like if [Qiskit/qiskit-aer#1686](https://github.com/Qiskit/qiskit-aer/issues/1686) were fixed we could make an all-in-one build? 

Yes, it's my hope, though I haven't followed Aer development for a while, and I could miss other subtle points / corner cases.

> So that would mean we would have to stop modifying the build string.

Yes, if we switch to an all-in-one build later.

> So
> 
> ```
> - run_constrained:
>   -  cuda-version <0.a0  # [cuda_compiler_version == "None"]
> ```
> 
> seems okay to me, assuming that installs the cpu version on a system without a gpu (rather than install the cuda libraries as well onto it).

In this case (when a conda env does not contain any CUDA packages, including the `cuda-version` metapackage) the CPU build would be installed. No other CUDA libraries would be installed along the way, because they are incompatible with this constraint. 

Conversely, if `cuda-version` is installed in an env, the GPU build would be picked by conda. This constraint acts as an extra safeguard to ensure conda behaves.

There's a subtlety, though. One can in principle install the GPU build on a system without GPUs too (ex: CI/CD). Unless we constrain also on `__cuda` (a virtual package representing the CUDA driver installed on the system), this is allowed.

> The build string approach seems okay as well. We can just live with the disruption when the all-in-one build becomes possible. 

Yes, if we stay with build string for now, we pay the one-time only cost later when switching. My assumption is @basnijholt prefers this option (the status quo), which I am fine with, as long as we all accept that there will be a disruption later.

> Will it default to the cpu version? I think that would be the best default, assuming conda has no way specify a preference, like install the gpu version if cuda-version is already installed. I would just want to avoid `conda install qiskit-aer` installing the gpu version on a system without a gpu.

So right now the default for `conda install qiskit-aer` is to install the GPU build (we bumped the build number in commit [d7227f4](https://github.com/conda-forge/qiskit-aer-feedstock/pull/19/commits/d7227f4fb608c6316d803a7f7f551b53cf8fedd2), which to conda/mamba means it's preferred during dependency resolution). If desired we can flip it.

**Comment by basnijholt** [comment]
> There's a subtlety, though. One can in principle install the GPU build on a system without GPUs too (ex: CI/CD).

This is actually what we do. We're installing a single lock file that contains CUDA packages, also on machines without a GPU, it would be great if this keeps working 😄 

So it seems like we can go ahead with merging in the PR in its current state?

**Comment by wshanks** [comment]
Can we match the build numbers so `conda install qiskit-aer` installs the CPU version? That should work for everyone but be less performant for users we a GPU. Otherwise won't users without a GPU end up with something that doesn't work?

Also, can we document the build variants? If nothing else, we can open and pin a "How to install GPU build" issue.

**Review by leofang — APPROVED** [review]

**Comment by basnijholt** [comment]
@wshanks I have now prioritized the CPU build and all CI jobs pass ✅

@leofang has approved too :tada:

Shall we merge?

**Comment by wshanks** [comment]
Sounds good! Those memory tricks might help the Windows build as well which I had trouble with (switching to an older vs runtime helped there). Now I just need to see if I can get macOS building supported upstream.

File: recipe/meta.yaml
Comment (jakirkham) [inline recipe/meta.yaml:47]: We may want to restrict this to CUDA 11. In CUDA 12, only `cuda-version` is added as via `run_exports` and it is a light metapackage used for alignment of CUDA Toolkit library versions (and does not add any bulky dependencies)

```suggestion
  ignore_run_exports_from:
    - {{ compiler('cuda') }}  # [(cuda_compiler_version or "").startswith("11")]
```

File: recipe/meta.yaml
Comment (jakirkham) [inline recipe/meta.yaml:45]: Would suggest just checking whether CUDA is used. That way it is simpler for folks to add CUDA on other architectures/platforms with fewer changes

```suggestion
    - CMAKE_BUILD_PARALLEL_LEVEL=1  # [cuda_compiler != "None"]
    - CUQUANTUM_ROOT=$PREFIX  # [cuda_compiler != "None"]
    - CUTENSOR_ROOT=$PREFIX  # [cuda_compiler != "None"]
    - AER_THRUST_BACKEND=CUDA  # [cuda_compiler != "None"]
    - AER_ENABLE_CUQUANTUM=ON  # [cuda_compiler != "None"]
```

File: recipe/meta.yaml
Comment (jakirkham) [inline recipe/meta.yaml:29]: Maybe this would work

```suggestion
    - export AER_CUDA_ARCH="7.0;7.5;8.0;8.6+PTX"  # [(cuda_compiler_version or "").startswith("11")]
```

File: recipe/meta.yaml
Comment (jakirkham) [inline recipe/meta.yaml:52]: Think we can simplify to...

```suggestion
    - {{ compiler('cuda') }}  # [cuda_compiler != "None"]
```

File: recipe/meta.yaml
Comment (jakirkham) [inline recipe/meta.yaml:72]: Think we can do this

```suggestion
    - cuda-version {{ cuda_compiler_version }}  # [cuda_compiler_version != "None"]
    # Qiskit Aer now uses both
    - custatevec                                # [cuda_compiler_version != "None"]
    - cutensornet                               # [cuda_compiler_version != "None"]
```

**Review by jakirkham — COMMENTED** [review]
Thanks Bas, Will, and Leo! 🙏

With an eye towards CUDA 12 and potentially other OSes or architectures if/when there is interest in adding them, think we can streamline a few things below to make that work easier (whenever that is pursued). Noted some items along these lines below

Mentioning these mainly for your consideration 🙂

**Comment by leofang** [comment]
Thanks, @jakirkham. Yes, I plan to make necessary updates when the migrators hit this feedstock.

File: recipe/meta.yaml
Comment (leofang) [inline recipe/meta.yaml:3]: FYI. I noticed we made a mistake here (not bumping the build number). Because we altered the build string in this PR, during uploading conda-build considered the CPU builds as distinct, and the new builds got uploaded. Since we're rebuilding in other PRs, this will be corrected.

File: recipe/meta.yaml
Comment (leofang) [reply recipe/meta.yaml:72]: Addressed in #21.

File: recipe/meta.yaml
Comment (leofang) [reply recipe/meta.yaml:52]: Addressed in #21.

File: recipe/meta.yaml
Comment (leofang) [reply recipe/meta.yaml:29]: Addressed in #21.

File: recipe/meta.yaml
Comment (leofang) [reply recipe/meta.yaml:45]: Addressed in #21.

File: recipe/meta.yaml
Comment (leofang) [reply recipe/meta.yaml:47]: Addressed in #21.

File: recipe/meta.yaml
Comment (wshanks) [reply recipe/meta.yaml:3]: I thought the changes here added 200 to the CPU build number to keep it higher than the GPU build and that we did want it uploaded with that new number?

Comment (leofang) [reply recipe/meta.yaml:3]: Yeah you're right, I forgot we switched to prioritize the cpu builds...

File: recipe/meta.yaml
Comment (leofang) [reply recipe/meta.yaml:71]: See https://github.com/conda-forge/qsimcirq-feedstock/pull/14.

**Pull request merged**
